Code/Framework/AzToolsFramework/AzToolsFramework/UI/Prefab/PrefabSaveLoadHandler.cpp

```diff
@@ -993,5 +1020,59 @@ namespace AzToolsFramework
         }
+
+        void PrefabSaveHandler::SourceFileRemoved(
+            AZStd::string relativePath, AZStd::string scanFolder, [[maybe_unused]] AZ::Uuid sourceUUID)
+        {
+            // This gets triggered for every source file. We only need source files that are prefabs and are loaded in the current level.
+            TemplateId loadedTemplateId = s_prefabSystemComponentInterface->GetTemplateIdFromFilePath(relativePath.c_str());
+            if (loadedTemplateId != InvalidTemplateId)
+            {
+                InstanceSetConstReference instancesMappedToTemplate =
+                    m_templateInstanceMapperInterface->FindInstancesOwnedByTemplate(loadedTemplateId);
+
+                if (!instancesMappedToTemplate.has_value() || instancesMappedToTemplate->get().size() == 0)
+                {
+                    // There are no instances mapping to this template. So it can be removed from memory.
+                    s_prefabSystemComponentInterface->RemoveTemplate(loadedTemplateId);
+                }
+                else
+                {
+                    TemplateId rootTemplateId = m_prefabEditorEntityOwnershipInterface->GetRootPrefabTemplateId();
+                    if (rootTemplateId == loadedTemplateId)
+                    {
+                        // Deletion of the current level that the user has opened needs to be handled differently. Skip for now
+                        return;
+                    }
+                    else
+                    {
+                        AzFramework::EntityContextId editorEntityContextId = AzFramework::EntityContextId::CreateNull();
+                        AzToolsFramework::EditorEntityContextRequestBus::BroadcastResult(
+                            editorEntityContextId, &AzToolsFramework::EditorEntityContextRequestBus::Events::GetEditorEntityContextId);
+
+                        if (m_prefabFocusPublicInterface->GetFocusedPrefabTemplateId(editorEntityContextId) == loadedTemplateId)
+                        {
+                            m_prefabFocusPublicInterface->FocusOnParentOfFocusedPrefab(editorEntityContextId);
+                        }
+
+                        auto button = QMessageBox::question(
+                            AzToolsFramework::GetActiveWindow(),
+                            QString(),
+                            QString("Prefab '%1' has been deleted from the file system. But we detected %2 instance(s) of it "
+                                    "in the level. Would you like to delete them ?")
+                                .arg(relativePath.c_str())
+                                .arg(instancesMappedToTemplate->get().size()),
+                            QMessageBox::Yes | QMessageBox::No);
+                        switch (button)
+                        {
+                        case QMessageBox::Yes:
+                            s_prefabSystemComponentInterface->RemoveTemplate(loadedTemplateId);
```

Comment:
Is this enough for removing nested instances in outliner? How does the nested instance get removed from outliner?

Reply:
Yes. Because RemoveTemplate() will also remove any links from the template to other dependent templates. I added an additional statement to trigger propagation on those dependent templates, which will eventually put the outliner in the right state : https://github.com/o3de/o3de/pull/14197/files#diff-ae6dbd26ad65eaa6de3465864f940853c6f5842fc3b9f68557b5b70600db1d48R1211